### PR TITLE
Add SPV_INTEL_variable_length_array extension

### PR DESCRIPTION
**Pull request by AlexeySotkin**
This is a preview extension specification, initial revision.

### CODE REVIEW

File: sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+== Validation Rules
+
+ - In control flow graph every
+<<VariableLengthArrayINTEL,*VariableLengthArrayINTEL*>> must be dominated by at
+least one <<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>>.
+
+ - In control flow graph every
+<<VariableLengthArrayINTEL,*VariableLengthArrayINTEL*>> must be postdominated by
+at least one <<OpSaveMemoryINTEL,*OpSaveMemoryINTEL*>>.
```
Comment (AlexeySachkov) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:200]: Should it be restore instead of save here?

Comment (AlexeySotkin) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc]: Good catch! Thanks.

**Comment by AlexeySotkin** [comment]
@jbrodman @mkinsner please review the PR.

**Review by AlexeySachkov — DISMISSED** [review]

File: sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+=== Instructions
+
+Modify Section 3.36.8. Memory Instructions, adding to the end of the list of
+instructions:
+
+[cols="1,1,3*3",width="100%"]
+|=====
+4+|[[OpVariableLengthArrayINTEL]]*OpVariableLengthArrayINTEL* +
+ +
+Allocate a runtime-sized array in the function memory, resulting in a pointer to
```
Comment (mkinsner) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:129]: Is "in the function memory" the definition that we want here?  Would it be better to say "function storage class"?

File: sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+it, which can be used with *OpLoad* and *OpStore*. The allocated memory is
+uninitialized. Loading from uninitialized memory causes undefined behavior. Size
```
Comment (mkinsner) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:131]: Should this be undefined behavior, or something weaker like "will return indeterminate values"?

Comment (AlexeySotkin) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:131]: "indeterminate values" sounds good to me.

File: sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc (hunk 2 above)
Comment (AlexeySotkin) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:129]: Right, I'm not sure about "function memory". What about:
"Allocate a runtime-sized array in memory, resulting in a pointer to it, which can be used with *OpLoad* and *OpStore*.
The memory is allocated in *Function* storage class and is uninitialized." ?

Comment (mkinsner) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:129]: Yes, I think that's an improvement and is fine with me.

File: sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc (hunk 3 above)
Comment (mkinsner) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_variable_length_array.asciidoc:131]: Sounds good

**Review by mkinsner — APPROVED** [review]

**Pull request merged**